### PR TITLE
feat: calculation

### PR DESCRIPTION
**Pull request by BardiyaFeili**
Hello,

Most launchers have a calculation functionality to quickly calculate equations and copy them to the clipboard.

My code uses the [rink-core](https://crates.io/crates/rink-core) crate to do the calculations and wl-copy to copy the result to the clipboard. wl-copy is available as wl-clipboard in most distros' repos.

You can enter an expression to calculate it's result. Pressing 'enter' on the result will copy it to the clipboard using the `wl-copy -t text/plain "result"` command.

This is my first pr so If there are any issues with the code please tell me so I can fix them.

### CODE REVIEW

**Review by nnyyxxxx — CHANGES_REQUESTED** [review]
needs to support xclip, needs to be togglable in the config file, needs a configuration.md change to accommodate that, and needs to be off by default

**Comment by BardiyaFeili** [comment]
What should the category for it in the configuration be called.
Is "Modes" good or something else?

**Comment by nnyyxxxx** [comment]
> What should the category for it in the configuration be called? Is "Modes" good or something else?

yes modes is fine, eventually we will add drun run and dmenu modes as well, so its perfect

**Comment by BardiyaFeili** [comment]
I've tried doing something like this and some other things but they don't work.

```rust
Command::new("sh")
    .arg("-c")
    .arg(format!("echo {} | xclip -i", &app.name))
    .spawn()
    .unwrap();
```
The reason is rust gives the arguments to sh without quotes but quotes are required for the pipe which is the only way I found to copy something to xclip

**Comment by abenz1267** [comment]
Using `evalexpr` is inferior to what other launchers do. Krunner, Walker, pop-launcher etc. all use `libqalculate`. In Rust you'd try to use `Rink`, as f.e. Anyrun is doing.

Anything else isn't adequate.

**Comment by abenz1267** [comment]
> I've tried doing something like this and some other things but they don't work.
> 
> ```rust
> Command::new("sh")
>     .arg("-c")
>     .arg(format!("echo {} | xclip -i", &app.name))
>     .spawn()
>     .unwrap();
> ```
> 
> The reason is rust gives the arguments to sh without quotes but quotes are required for the pipe which is the only way I found to copy something to xclip

https://doc.rust-lang.org/std/process/index.html#handling-io

**Comment by BardiyaFeili** [comment]
> Using `evalexpr` is inferior to what other launchers do. Krunner, Walker, pop-launcher etc. all use `libqalculate`. In Rust you'd try to use `Rink`, as f.e. Anyrun is doing.
> 
> Anything else isn't adequate.

I had a look at it and it looks like it will be better than the evalexpr crate I was using. 

I will switch to it tomorrow.

**Comment by abenz1267** [comment]
What's the state here? Last commit was 4 days ago.

**Comment by BardiyaFeili** [comment]
> What's the state here? Last commit was 2 days ago.

I have been busy for a few days. I think that I will have time to work on this on Sunday.

**Comment by abenz1267** [comment]
> > What's the state here? Last commit was 2 days ago.
> 
> I have been busy for a few days. I think that I will have time to work on this on Sunday.

I just wasn't sure, because the last thing you wrote was about switching to Rink, which i thought happened in the last commit. No worries.

**Comment by BardiyaFeili** [comment]
> I just wasn't sure, because the last thing you wrote was about switching to Rink, which i thought happened in the last commit. No worries.

The switch to rink is complete and the calculator if working fine right now, The only thing left is supporting xclip.

**Comment by abenz1267** [comment]
> The implementation may be harder, but I was wondering maybe using qalculator may be a better option as it has a better OOB experience with more built in functions and units? Just a thought
> 
> The CLI "qcalc" seems to be a good starting point, taking in the input and outputting its interpretation and the answer. And if the user selects the result to copy, copy-friendly version can be obtained with `qcalc -t`

The implementation isn't really that much harder. Personally using rink for simplicity... is ok. But using libqalculate would be better. That's what Krunner, Walker and other runners also user. So people are already familiar with that.

**Comment by BardiyaFeili** [comment]
> conflicts need fixing, due to: #114

Fixed all of them

**Comment by BardiyaFeili** [comment]
Although building the app is currently failing because of the changes that were made to the code base.

**Comment by BardiyaFeili** [comment]
The calculator is working again, There are still some bugs and problems which I will fix as soon as I can.

**Comment by nnyyxxxx** [comment]
more conflicts

File: configuration.md [hunk 1]
```diff
@@ -186,9 +197,14 @@ When using custom CSS (style.css), you can also customize the error overlay appe
...
 
 ### Debug Options
+
 - `disable_auto_focus`: Prevents the window from automatically holding all input
 - `enable_logging`: Enables logging to the terminal window Hyprlauncher was launched from
 
+### Modes
+
+- calculator: Do calculations when starting your search with '='
+
```
Comment (nnyyxxxx) [inline configuration.md:207]: This doesn't look correct, header should be "Calculator Mode". More documentation is needed as well, like the modes below.

**Review by nnyyxxxx — CHANGES_REQUESTED** [review]

File: src/search.rs [hunk 2]
```diff
@@ -218,6 +222,13 @@ pub async fn search_applications(
                     }
                 }
 
+                if results.is_empty()
+                    && calculator_enabled
+                    && query.trim().chars().next().unwrap().is_ascii_digit()
+                {
+                    results.push(create_calc_entry(&query));
```
Comment (nnyyxxxx) [inline src/search.rs:229]: this is only checking if the first character is a digit, PR description is misleading.

File: src/search.rs [hunk 3]
```diff
@@ -20,6 +21,7 @@ const BONUS_SCORE_BINARY: i64 = 3000;
 const BONUS_SCORE_KEYWORD_MATCH: i64 = 2500;
 const BONUS_SCORE_CATEGORY_MATCH: i64 = 2000;
 const BONUS_SCORE_WEB_SEARCH: i64 = -1000;
+const BONUS_SCORE_CALC: i64 = 3000;
```
Comment (nnyyxxxx) [inline src/search.rs:24]: ``BONUS_SCORE_CALC`` is not being used properly, see ``=`` review below

**Review by nnyyxxxx — CHANGES_REQUESTED** [review]

File: src/search.rs [hunk 4]
```diff
@@ -450,3 +463,46 @@ fn load_history() -> HashMap<String, HistoryEntry> {
...
+
+    let res = handle_calculation(query);
+
+    SearchResult {
+        app: AppEntry {
+            name: res.clone(),
+            description: String::from("Copy to clipboard"),
+            path: String::new(),
+            exec: format!("wl-copy -t text/plain \"{}\"", res),
+            icon_name: String::from("calculator"),
```
Comment (nnyyxxxx) [inline src/search.rs:481]:
```suggestion
            icon_name: String::from("accessories-calculator"),
```

**Review by nnyyxxxx — CHANGES_REQUESTED** [review]

File: src/search.rs (hunk 4 above)
Comment (nnyyxxxx) [reply src/search.rs:481]: this fixes the icon issue

Comment (nnyyxxxx) [reply src/search.rs:481]: image

Comment (nnyyxxxx) [reply src/search.rs:481]: why was this marked as resolved?

File: src/search.rs (hunk 2 above)
Comment (BardiyaFeili) [reply src/search.rs:229]: I have changed the way how entering the calculator mode works. I will also change the documentation accordingly.

**Review by nnyyxxxx — APPROVED** [review]
lgtm

@adamperkowski review

**Review by nnyyxxxx — CHANGES_REQUESTED** [review]
https://github.com/hyprutils/hyprlauncher/actions/runs/12169684976/job/33944238886?pr=88

fix spelling mistakes

**Review by nnyyxxxx — CHANGES_REQUESTED** [review]
https://github.com/hyprutils/hyprlauncher/commit/5a378f4ef6849716033ed5bd558eba655a68d5be

with this commit, the new category will have to get added to ``let required_categories``

**Review by nnyyxxxx — APPROVED** [review]

**Pull request merged**